internal/providers/pke/pkeworkflow/create_cluster.go

```diff
@@ -123,6 +123,28 @@ func (w CreateClusterWorkflow) Execute(ctx workflow.Context, input CreateCluster
 		}
 	}
 
+	// Select images (if not specified)
+	{
+		activityInput := SelectImagesActivityInput{ClusterID: input.ClusterID, NodePools: nodePools}
+		err := workflow.ExecuteActivity(ctx, SelectImagesActivityName, activityInput).Get(ctx, &nodePools)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
TBH I liked the idea of accepting full specification in the node pool create activities, but this is not really a nice solution. I think this activity should be called separately for each node node pool (in case it has no image defined) and should return an image only.

Reply:
I tried this approach too, but it made the workflow too complicated with all the glue code, iterations, conversions between lists and sets (with a stable/consistent order) etc. Also it would need even more workflow code (maintained twice: both in the create and update flow).

Reply:
Common code can be extracted to functions and reused in workflows.

Reply:
Original comment fixed in 4d73fbd01194366cda3af797608367dfab1cfb4d.
Secondary comment fixed in e43d6fd3eedd4d530de6565a1a1150a30e56b9f1.